io/private/stac/Catalog.cpp

```diff
@@ -0,0 +1,221 @@
+            const std::string linkPath = m_utils.stacValue<std::string>(
+                link, "href", m_json);
+
+            const std::string absLinkPath = m_utils.handleRelativePath(m_path, linkPath);
+
+            m_pool.add([this, linkType, absLinkPath, filters, rawReaderArgs]()
+            {
+                try
+                {
+                    if (linkType == "item")
+                    {
+                        NL::json itemJson = m_connector.getJson(absLinkPath);
+                        Item item(itemJson, absLinkPath, m_connector, m_validate);
+
+                        bool valid = item.init(*filters.itemFilters,
+                            rawReaderArgs, m_schemaUrls);
+                        if (valid)
+                        {
+                            std::lock_guard<std::mutex> lock(m_mutex);
+                            m_itemList.push_back(item);
+                        }
+                    }
+                    else if (linkType == "catalog")
+                    {
+                        NL::json catalogJson = m_connector.getJson(absLinkPath);
+                        std::unique_ptr<Catalog> catalog(new Catalog(
+                            catalogJson, absLinkPath, m_connector, m_pool,
+                            m_validate));
+
+                        bool valid = catalog->init(filters, rawReaderArgs,
+                            m_schemaUrls);
+                        if (valid)
+                        {
+                            std::lock_guard<std::mutex> lock(m_mutex);
+                            m_subCatalogs.push_back(std::move(catalog));
+                        }
+                    }
+                    else if (linkType == "collection")
+                    {
+                        NL::json collectionJson = m_connector.getJson(absLinkPath);
+                        std::unique_ptr<Collection> collection(new Collection(
+                            collectionJson, absLinkPath, m_connector, m_pool,
+                            m_validate));
+
+                        bool valid = collection->init(filters, rawReaderArgs,
+                            m_schemaUrls);
+                        if (valid)
+                        {
+                            std::lock_guard<std::mutex> lock(m_mutex);
+                            m_subCatalogs.push_back(std::move(collection));
+                        }
+                    }
+                }
+                catch (std::exception& e)
+                {
+                    std::lock_guard<std::mutex> lock(m_mutex);
+                    StacError p {absLinkPath, e.what()};
+                    m_errors.push_back(p);
+                }
+                catch (...)
+                {
+                    std::lock_guard<std::mutex> lock(m_mutex);
+                    StacError p {absLinkPath, "Unknown error"};
+                    m_errors.push_back(p);
+                }
+            });
+        }
+
+
+        if (isRoot)
```

Comment:
`isRoot` would only be called as true from the `StacReader`, since that's the entry point into the STAC structure, if there are sub-catalogs or sub-collections they'll be created with `isRoot=false` like [here](https://github.com/PDAL/PDAL/blob/relative-stac/io/private/stac/Catalog.cpp#L105-L106). So we'll only be doing a join if we're handling the original collection/catalog that was created by `StacReader`